cql3/statements/ks_prop_defs.cc

```diff
@@ -63,6 +68,63 @@ static locator::replication_strategy_config_options prepare_options(
         }
     }
 
+    auto expand_dc_racks = [&] (const sstring& dc, const locator::replication_strategy_config_option& rf) {
+        logger.debug("expand_dc_racks: dc={} rf={} all_dcs={}", dc, rf, all_dcs);
+
+        std::unordered_set<sstring> allowed_racks;
+        std::vector<sstring> dc_racks;
+        auto it = all_dcs.find(dc);
+        if (it != all_dcs.end()) {
+            dc_racks = it->second | std::views::keys | std::ranges::to<std::vector<sstring>>();
+            allowed_racks = std::ranges::to<std::unordered_set<sstring>>(dc_racks);
+        } else if (!tm.get_topology().get_datacenters().contains(dc)) {
+            throw exceptions::configuration_exception(fmt::format("Unrecognized datacenter name '{}'", dc));
+        }
+
+        auto& topo = tm.get_topology();
+        auto normal_nodes = [&] (const sstring& rack) {
+            int count = 0;
+            for (auto n : it->second.at(rack)) {
+                count += int(topo.get_node(n).is_normal());
+            }
+            return count;
+        };
+
+        auto data = locator::abstract_replication_strategy::parse_replication_factor(rf, allowed_racks);
+
+        if (data.is_rack_based()) {
+            options[dc] = data.get_rack_list();
+        } else if (data.count() == 0) {
+            options.emplace(dc, "0");
+        } else if (force_racks) {
+            if (data.count() > dc_racks.size()) {
+                throw exceptions::configuration_exception(fmt::format(
+                        "Replication factor {} exceeds the number of racks ({}) in dc {}", data.count(), dc_racks.size(), dc));
+            }
+            if (old_options.count(dc)) {
+                auto& old_rf_val = old_options.at(dc);
+                auto old_rf = locator::abstract_replication_strategy::parse_replication_factor(old_rf_val, allowed_racks);
+                if (old_rf.is_rack_based() && old_rf.count() == data.count()) {
+                    options[dc] = old_rf_val;
+                    return;
+                }
+            }
+            // If the replication factor is less than the number of racks, pick rf racks at random.
+            if (data.count() < dc_racks.size()) {
+                static thread_local auto gen = std::default_random_engine(std::random_device{}());
+                std::ranges::shuffle(dc_racks, gen);
+                // Prefer placing into racks which have more normal nodes.
+                std::sort(dc_racks.begin(), dc_racks.end(), [&](const sstring& a, const sstring& b) {
```

Comment:
Is choosing racks which have the most nodes much better than completely at random? With the current approach, a rack with the most nodes will have tablets from all keyspaces and a rack with the least nodes won't get any tablets. I assume we'll usually have all racks with the same number of nodes, so these won't be issues but in that case a random choice is still no worse.
Node count may also be misleading in clusters with different node sizes.
If we want to complicate things, we could take a random sample from racks, weighing them by their node counts
And if we wanted to complicate this even further, we could look at the storage utilization of each rack and choose the least utilized ones.

Reply:
Auto expansion is pretty much for tests only. In production, you don't want random placement because capacity needs to be provided accordingly and planned. So I don't think it matters.

The primary goal here is to avoid choosing racks which don't have normal nodes. Such nodes will be placed at the back.